### PR TITLE
Update Kani to Rust nightly-2022-10-11

### LINKED ISSUE #1746: Update rust toolchain version for Kani 0.13
We should update the rust toolchain version once every release. For 0.13, we will try to update to version nightly-2022-10-11

### PR DESCRIPTION
**Pull request by tedinski**
### Description of changes: 

Updates to newer nightly. Notable impacting changes from upstream:

1. https://github.com/rust-lang/rust/pull/102675
   * `CastKind::Misc` became `CastKind::IntToInt | CastKind::FloatToFloat | CastKind::FloatToInt | CastKind::IntToFloat | CastKind::FnPtrToPtr | CastKind::PtrToPtr`

2. https://github.com/rust-lang/rust/pull/98588
   * MIR created `ConstantKind::Unevaluated` to replace `ContantKind::Ty` wrapping a `ConstKind::Unevaluated`. Note the `ANT` in `Constant` distinguishing these!
   * `visit_const` was removed from `MirVisitor` as a consequence

3. https://github.com/rust-lang/rust/pull/101212
   * Initial implementation of `dyn*` might create problems for us in the future but simple change for now
   * It looks like they just littered fixmes about... so the only place that broke I've added a `codegen_unimplemented` for now

4. https://github.com/rust-lang/rust/pull/98582
   * Reintroduce `ProjectionElem::OpaqueCast`
   * We'd seen this before (thanks, @celinval!). I re-added the code from https://github.com/model-checking/kani/pull/1399

5. https://github.com/rust-lang/rust/pull/101061
   * Seems like `intrinsics::assert_uninit_valid::<bool>()` is ok now? Switch to ! like that pull request

### Resolved issues:

Resolves #1746 

### Call-outs:

* `simd_shuffle` is disabled right now, but there's probably a new issue in the way of re-enabling it (it is or can be plain `simd_shuffle` now not just e.g. `simd_shuffle8`) @adpaco-aws 
* `let_else` is getting stabilized? Woo hoo!

### Testing:

* How is this change tested? ci

* Is this a refactor change? no

### Checklist
- [x] Each commit message has a non-empty body, explaining why the change was made
- [x] Methods or procedures are documented
- [x] Regression or unit tests are included, or existing tests cover the modified code
- [x] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: kani-compiler/src/codegen_cprover_gotoc/codegen/operand.rs [hunk 1]
```diff
@@ -47,28 +47,37 @@ impl<'tcx> GotocCtx<'tcx> {
...
+    }
 
-        self.codegen_const(const_, Some(&c.span))
+    fn codegen_const_unevaluated(
+        &mut self,
+        unevaluated: UnevaluatedConst<'tcx>,
+        ty: Ty<'tcx>,
+        span: Option<&Span>,
+    ) -> Expr {
+        debug!("The literal was a Unevaluated");
```
Comment (celinval) [inline kani-compiler/src/codegen_cprover_gotoc/codegen/operand.rs:66]: Can you please add the constant?

```rust
debug!(?unevaluated, "codegen_const_unevaluated");
```

Nit: I usually add the name of the function to the log so it's easy to find it in the code. But I don't think we have any guidelines on this. So just a suggestion.

File: kani-compiler/src/codegen_cprover_gotoc/codegen/operand.rs [hunk 2]
```diff
@@ -47,28 +47,37 @@ impl<'tcx> GotocCtx<'tcx> {
...
-                debug!("The literal was a Unevaluated");
-                let const_val = self
-                    .tcx
-                    .const_eval_resolve(ty::ParamEnv::reveal_all(), unevaluated, None)
-                    .unwrap();
-                self.codegen_const_value(const_val, lit.ty(), span)
-            }
+            // A `ConstantKind::Ty(ConstKind::Unevaluated)` should no longer show up
+            // and should be a `ConstantKind::Unevaluated` instead (and thus handled
+            // in `codegen_constant` not here.)
```
Comment (celinval) [inline kani-compiler/src/codegen_cprover_gotoc/codegen/operand.rs:79]: in `codegen_const_unevaluated`

File: tests/kani/ProjectionElem/OpaqueCast/check_opaque_cast.rs [hunk 3]
```diff
@@ -2,10 +2,6 @@
 //
 // Modifications Copyright Kani Contributors
 // See GitHub history for details.
-//
-// NOTE: The initial fix for this has been reverted from rustc. I'm keeping this test here so we
```
Comment (celinval) [inline tests/kani/ProjectionElem/OpaqueCast/check_opaque_cast.rs:6]: Hopefully this was useful. :)

**Review by celinval — APPROVED** [review]
Awesome! Just some nits.

File: kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs [hunk 4]
```diff
@@ -404,8 +404,11 @@ impl<'tcx> GotocCtx<'tcx> {
             }};
         }
 
-        if let Some(stripped) = intrinsic.strip_prefix("simd_shuffle") {
-            let _n: u64 = stripped.parse().unwrap();
+        if let Some(_stripped) = intrinsic.strip_prefix("simd_shuffle") {
+            // TODO: can be empty now (i.e. `simd_shuffle` instead of `simd_shuffle8`)
+            // `parse` fails on empty, so comment that bit of code out.
+            // To re-enable this we'll need to investigate how size is computed now.
+            // let n: u64 = stripped.parse().unwrap();
```
Comment (adpaco-aws) [inline kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs:411]: Thanks for the heads-up! I'll figure this out as part of the work being done in #1148.

File: tests/kani/Intrinsics/Assert/uninit_valid_panic.rs [hunk 5]
```diff
@@ -5,11 +5,11 @@
...
 use std::intrinsics;
 
-// The code below attempts to leave type `bool` uninitialized, causing the
+// The code below attempts to leave type `!` uninitialized, causing the
 // intrinsic `assert_uninit_valid` to generate a panic during compilation.
 #[kani::proof]
 fn main() {
     let _var: () = unsafe {
-        intrinsics::assert_uninit_valid::<bool>();
+        intrinsics::assert_uninit_valid::<!>();
```
Comment (adpaco-aws) [inline tests/kani/Intrinsics/Assert/uninit_valid_panic.rs:13]: The problem is that this triggers the same path as `tests/kani/Intrinsics/Assert/inhabited_panic.rs`. Can you check that you get the message below if you replace `bool` with `&i32`?
```
attempted to leave type `&i32` uninitialized, which is invalid
```

**Review by adpaco-aws — CHANGES_REQUESTED** [review]

File: tests/kani/Intrinsics/Assert/uninit_valid_panic.rs (hunk 5 above)
Comment (tedinski) [reply tests/kani/Intrinsics/Assert/uninit_valid_panic.rs]: Yep! Fixed.

**Review by adpaco-aws — APPROVED** [review]

**Pull request merged**